File: test2/src/main/java/xyz/yggdrazil/delaunay/voronoi/nodename/as3delaunay/Site.java
package xyz.yggdrazil.delaunay.voronoi.nodename.as3delaunay;

import xyz.yggdrazil.delaunay.geom.Point;
import xyz.yggdrazil.delaunay.geom.Rectangle;

import java.awt.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Stack;

public final class Site implements ICoord {

    final private static double EPSILON = .005;
    private static Stack<Site> _pool = new Stack();
    public Color color;
    public double weight;
    // the edges that define this Site's Voronoi region:
    public ArrayList<Edge> edges;
    private Point coord;
    private int _siteIndex;
    // which end of each edge hooks up with the previous edge in edges:
    private ArrayList<LR> _edgeOrientations;
    // ordered list of points that define the region clipped to bounds:
    private ArrayList<Point> _region;

    public Site(Point p, int index, double weight, Color color) {
        init(p, index, weight, color);
    }

    public static Site create(Point p, int index, double weight, Color color) {
        if (_pool.size() > 0) {
            return _pool.pop().init(p, index, weight, color);
        } else {
            return new Site(p, index, weight, color);
        }
    }

    public static void sortSites(ArrayList<Site> sites) {
        //sites.sort(Site.compare);
        Collections.sort(sites, new Comparator<Site>() {
            @Override
            public int compare(Site o1, Site o2) {
                return (int) Site.compare(o1, o2);
            }
        });
    }

    /**
     * sort sites on y, then x, coord also change each site's _siteIndex to
     * match its new position in the list so the _siteIndex can be used to
     * identify the site for nearest-neighbor queries
     * <p>
     * haha "also" - means more than one responsibility...
     */
    private static double compare(Site s1, Site s2) {
        int returnValue = Voronoi.compareByYThenX(s1, s2);

        // swap _siteIndex values if necessary to match new ordering:
        int tempIndex;
        if (returnValue == -1) {
            if (s1._siteIndex > s2._siteIndex) {
                tempIndex = s1._siteIndex;
                s1._siteIndex = s2._siteIndex;
                s2._siteIndex = tempIndex;
            }
        } else if (returnValue == 1) {
            if (s2._siteIndex > s1._siteIndex) {
                tempIndex = s2._siteIndex;
                s2._siteIndex = s1._siteIndex;
                s1._siteIndex = tempIndex;
            }

        }

        return returnValue;
    }

    private static boolean closeEnough(Point p0, Point p1) {
        return Point.Companion.distance(p0, p1) < EPSILON;
    }

    @Override
    public Point getCoord() {
        return coord;
    }

    private Site init(Point p, int index, double weight, Color color) {
        coord = p;
        _siteIndex = index;
        this.weight = weight;
        this.color = color;
        edges = new ArrayList();
        _region = null;
        return this;
    }

    @Override
    public String toString() {
        return "Site " + _siteIndex + ": " + getCoord();
    }

    private void move(Point p) {
        clear();
        coord = p;
    }

    public void dispose() {
        coord = null;
        clear();
        _pool.push(this);
    }

    private void clear() {
        if (edges != null) {
            edges.clear();
            edges = null;
        }
        if (_edgeOrientations != null) {
            _edgeOrientations.clear();
            _edgeOrientations = null;
        }
        if (_region != null) {
            _region.clear();
            _region = null;
        }
    }

    void addEdge(Edge edge) {
        edges.add(edge);
    }

    public Edge nearestEdge() {
        // edges.sort(Edge.compareSitesDistances);
        Collections.sort(edges, new Comparator<Edge>() {
            @Override
            public int compare(Edge o1, Edge o2) {
                return (int) Edge.Companion.compareSitesDistances(o1, o2);
            }
        });
        return edges.get(0);
    }

    ArrayList<Site> neighborSites() {
        if (edges == null || edges.isEmpty()) {
            return new ArrayList();
        }
        if (_edgeOrientations == null) {
            reorderEdges();
        }
        ArrayList<Site> list = new ArrayList();
        for (Edge edge : edges) {
            list.add(neighborSite(edge));
        }
        return list;
    }

    private Site neighborSite(Edge edge) {
        if (this == edge.getLeftSite()) {
            return edge.getRightSite();
        }
        if (this == edge.getRightSite()) {
            return edge.getLeftSite();
        }
        return null;
    }

    ArrayList<Point> region(Rectangle clippingBounds) {
        if (edges == null || edges.isEmpty()) {
            return new ArrayList();
        }
        if (_edgeOrientations == null) {
            reorderEdges();
            _region = clipToBounds(clippingBounds);
            if ((new Polygon(_region)).winding() == Winding.CLOCKWISE) {
                Collections.reverse(_region);
            }
        }
        return _region;
    }

    private void reorderEdges() {
        //trace("edges:", edges);
        EdgeReorderer reorderer = new EdgeReorderer(edges, Vertex.class);
        edges = reorderer.getEdges();
        //trace("reordered:", edges);
        _edgeOrientations = reorderer.getEdgeOrientations();
        reorderer.dispose();
    }

    private ArrayList<Point> clipToBounds(Rectangle bounds) {
        ArrayList<Point> points = new ArrayList();
        int n = edges.size();
        int i = 0;
        Edge edge;
        while (i < n && (!edges.get(i).getVisible())) {
            ++i;
        }

        if (i == n) {
            // no edges visible
            return new ArrayList();
        }
        edge = edges.get(i);
        LR orientation = _edgeOrientations.get(i);
        points.add(edge.getClippedEnds().get(orientation));
        points.add(edge.getClippedEnds().get((LR.Companion.other(orientation))));

        for (int j = i + 1; j < n; ++j) {
            edge = edges.get(j);
            if (!edge.getVisible()) {
                continue;
            }
            connect(points, j, bounds, false);
        }
        // close up the polygon by adding another corner point of the bounds if needed:
        connect(points, i, bounds, true);

        return points;
    }

    private void connect(ArrayList<Point> points, int j, Rectangle bounds, boolean closingUp) {
        Point rightPoint = points.get(points.size() - 1);
        Edge newEdge = edges.get(j);
        LR newOrientation = _edgeOrientations.get(j);
        // the point that  must be connected to rightPoint:
        Point newPoint = newEdge.getClippedEnds().get(newOrientation);
        if (!closeEnough(rightPoint, newPoint)) {
            // The points do not coincide, so they must have been clipped at the bounds;
            // see if they are on the same border of the bounds:
            if (rightPoint.getX() != newPoint.getX()
                    && rightPoint.getY() != newPoint.getY()) {
                // They are on different borders of the bounds;
                // insert one or two corners of bounds as needed to hook them up:
                // (NOTE this will not be correct if the region should take up more than
                // half of the bounds rect, for then we will have gone the wrong way
                // around the bounds and included the smaller part rather than the larger)
                int rightCheck = BoundsCheck.check(rightPoint, bounds);
                int newCheck = BoundsCheck.check(newPoint, bounds);
                double px, py;
                if ((rightCheck & BoundsCheck.RIGHT) != 0) {
                    px = bounds.getRight();
                    if ((newCheck & BoundsCheck.BOTTOM) != 0) {
                        py = bounds.getBottom();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.TOP) != 0) {
                        py = bounds.getTop();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.LEFT) != 0) {
                        if (rightPoint.getY() - bounds.getY() + newPoint.getY() - bounds.getY() < bounds.getHeight()) {
                            py = bounds.getTop();
                        } else {
                            py = bounds.getBottom();
                        }
                        points.add(new Point(px, py));
                        points.add(new Point(bounds.getLeft(), py));
                    }
                } else if ((rightCheck & BoundsCheck.LEFT) != 0) {
                    px = bounds.getLeft();
                    if ((newCheck & BoundsCheck.BOTTOM) != 0) {
                        py = bounds.getBottom();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.TOP) != 0) {
                        py = bounds.getTop();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.RIGHT) != 0) {
                        if (rightPoint.getY() - bounds.getY() + newPoint.getY() - bounds.getY() < bounds.getHeight()) {
                            py = bounds.getTop();
                        } else {
                            py = bounds.getBottom();
                        }
                        points.add(new Point(px, py));
                        points.add(new Point(bounds.getRight(), py));
                    }
                } else if ((rightCheck & BoundsCheck.TOP) != 0) {
                    py = bounds.getTop();
                    if ((newCheck & BoundsCheck.RIGHT) != 0) {
                        px = bounds.getRight();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.LEFT) != 0) {
                        px = bounds.getLeft();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.BOTTOM) != 0) {
                        if (rightPoint.getX() - bounds.getX() + newPoint.getX() - bounds.getX() < bounds.getWidth()) {
                            px = bounds.getLeft();
                        } else {
                            px = bounds.getRight();
                        }
                        points.add(new Point(px, py));
                        points.add(new Point(px, bounds.getBottom()));
                    }
                } else if ((rightCheck & BoundsCheck.BOTTOM) != 0) {
                    py = bounds.getBottom();
                    if ((newCheck & BoundsCheck.RIGHT) != 0) {
                        px = bounds.getRight();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.LEFT) != 0) {
                        px = bounds.getLeft();
                        points.add(new Point(px, py));
                    } else if ((newCheck & BoundsCheck.TOP) != 0) {
                        if (rightPoint.getX() - bounds.getX() + newPoint.getX() - bounds.getX() < bounds.getWidth()) {
                            px = bounds.getLeft();
                        } else {
                            px = bounds.getRight();
                        }
                        points.add(new Point(px, py));
                        points.add(new Point(px, bounds.getTop()));
                    }
                }
            }
            if (closingUp) {
                // newEdge's ends have already been added
                return;
            }
            points.add(newPoint);
        }
        Point newRightPoint = newEdge.getClippedEnds().get(LR.Companion.other(newOrientation));
        if (!closeEnough(points.get(0), newRightPoint)) {
            points.add(newRightPoint);
        }
    }

    public double get_x() {
        return coord.getX();
    }

    public double get_y() {
        return coord.getY();
    }

    public double dist(ICoord p) {
        return Point.Companion.distance(p.getCoord(), this.coord);
    }
}

final class BoundsCheck {

    final public static int TOP = 1;
    final public static int BOTTOM = 2;
    final public static int LEFT = 4;
    final public static int RIGHT = 8;

    /**
     * @param point
     * @param bounds
     * @return an int with the appropriate bits set if the Point lies on the
     * corresponding bounds lines
     */
    public static int check(Point point, Rectangle bounds) {
        int value = 0;
        if (point.getX() == bounds.getLeft()) {
            value |= LEFT;
        }
        if (point.getX() == bounds.getRight()) {
            value |= RIGHT;
        }
        if (point.getY() == bounds.getTop()) {
            value |= TOP;
        }
        if (point.getY() == bounds.getBottom()) {
            value |= BOTTOM;
        }
        return value;
    }

}

File: test2/src/main/java/xyz/yggdrazil/delaunay/voronoi/nodename/as3delaunay/Site.kt
package xyz.yggdrazil.delaunay.voronoi.nodename.as3delaunay

import xyz.yggdrazil.delaunay.geom.Point
import xyz.yggdrazil.delaunay.geom.Rectangle

import java.awt.*
import java.util.ArrayList
import java.util.Collections
import java.util.Comparator
import java.util.Stack

class Site(p: Point, index: Int, weight: Double, var color: Color?) : ICoord {

    override var coord: Point? = null
    var weight: Double = 0.toDouble()
    // the edges that define this Site's Voronoi region:
    var edges: ArrayList<Edge>? = null

    private var _siteIndex: Int = 0
    // which end of each edge hooks up with the previous edge in edges:
    private var _edgeOrientations: ArrayList<LR>? = null
    // ordered list of points that define the region clipped to bounds:
    private var _region: ArrayList<Point>? = null

    init {
        init(p, index, weight, color)
    }

    private fun init(p: Point, index: Int, weight: Double, color: Color?): Site {
        coord = p
        _siteIndex = index
        this.weight = weight
        this.color = color
        edges = ArrayList()
        _region = null
        return this
    }

    override fun toString(): String {
        return "Site $_siteIndex: $coord"
    }

    private fun move(p: Point) {
        clear()
        coord = p
    }

    fun dispose() {
        coord = null
        clear()
        pool.push(this)
    }

    private fun clear() {
        if (edges != null) {
            edges!!.clear()
            edges = null
        }
        if (_edgeOrientations != null) {
            _edgeOrientations!!.clear()
            _edgeOrientations = null
        }
        if (_region != null) {
            _region!!.clear()
            _region = null
        }
    }

    internal fun addEdge(edge: Edge) {
        edges!!.add(edge)
    }

    fun nearestEdge(): Edge {
        // edges.sort(Edge.compareSitesDistances);
        Collections.sort(edges!!) { o1, o2 -> Edge.compareSitesDistances(o1, o2).toInt() }
        return edges!![0]
    }

    internal fun neighborSites(): ArrayList<Site?> {
        if (edges == null || edges!!.isEmpty()) {
            return ArrayList()
        }
        if (_edgeOrientations == null) {
            reorderEdges()
        }
        val list = ArrayList<Site?>()
        for (edge in edges!!) {
            list.add(neighborSite(edge))
        }
        return list
    }

    private fun neighborSite(edge: Edge): Site? {
        if (this == edge.leftSite) {
            return edge.rightSite
        }
        if (this == edge.rightSite) {
            return edge.leftSite
        }
        return null
    }

    internal fun region(clippingBounds: Rectangle): ArrayList<Point>? {
        if (edges == null || edges!!.isEmpty()) {
            return ArrayList()
        }
        if (_edgeOrientations == null) {
            reorderEdges()
            _region = clipToBounds(clippingBounds)
            if (Polygon(_region!!).winding() == Winding.CLOCKWISE) {
                Collections.reverse(_region!!)
            }
        }
        return _region
    }

    private fun reorderEdges() {
        //trace("edges:", edges);
        val reorderer = EdgeReorderer(edges!!, Vertex::class.java)
        edges = reorderer.edges
        //trace("reordered:", edges);
        _edgeOrientations = reorderer.edgeOrientations
        reorderer.dispose()
    }

    private fun clipToBounds(bounds: Rectangle): ArrayList<Point> {
        val points = ArrayList<Point>()
        val n = edges!!.size
        var i = 0
        var edge: Edge
        while (i < n && !edges!![i].visible) {
            ++i
        }

        if (i == n) {
            // no edges visible
            return ArrayList()
        }
        edge = edges!![i]
        val orientation = _edgeOrientations!![i]
        points.add(edge.clippedEnds!![orientation]!!)
        points.add(edge.clippedEnds!![LR.other(orientation)]!!)

        for (j in i + 1..n - 1) {
            edge = edges!![j]
            if (!edge.visible) {
                continue
            }
            connect(points, j, bounds, false)
        }
        // close up the polygon by adding another corner point of the bounds if needed:
        connect(points, i, bounds, true)

        return points
    }

    private fun connect(points: ArrayList<Point>, j: Int, bounds: Rectangle, closingUp: Boolean) {
        val rightPoint = points[points.size - 1]
        val newEdge = edges!![j]
        val newOrientation = _edgeOrientations!![j]
        // the point that  must be connected to rightPoint:
        val newPoint = newEdge.clippedEnds!![newOrientation]
        if (!closeEnough(rightPoint, newPoint!!)) {
            // The points do not coincide, so they must have been clipped at the bounds;
            // see if they are on the same border of the bounds:
            if (rightPoint.x != newPoint.x && rightPoint.y != newPoint.y) {
                // They are on different borders of the bounds;
                // insert one or two corners of bounds as needed to hook them up:
                // (NOTE this will not be correct if the region should take up more than
                // half of the bounds rect, for then we will have gone the wrong way
                // around the bounds and included the smaller part rather than the larger)
                val rightCheck = BoundsCheck.check(rightPoint, bounds)
                val newCheck = BoundsCheck.check(newPoint!!, bounds)
                val px: Double
                val py: Double
                if (rightCheck and BoundsCheck.RIGHT != 0) {
                    px = bounds.right
                    if (newCheck and BoundsCheck.BOTTOM != 0) {
                        py = bounds.bottom
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.TOP != 0) {
                        py = bounds.top
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.LEFT != 0) {
                        if (rightPoint.y - bounds.y + newPoint.y - bounds.y < bounds.height) {
                            py = bounds.top
                        } else {
                            py = bounds.bottom
                        }
                        points.add(Point(px, py))
                        points.add(Point(bounds.left, py))
                    }
                } else if (rightCheck and BoundsCheck.LEFT != 0) {
                    px = bounds.left
                    if (newCheck and BoundsCheck.BOTTOM != 0) {
                        py = bounds.bottom
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.TOP != 0) {
                        py = bounds.top
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.RIGHT != 0) {
                        if (rightPoint.y - bounds.y + newPoint.y - bounds.y < bounds.height) {
                            py = bounds.top
                        } else {
                            py = bounds.bottom
                        }
                        points.add(Point(px, py))
                        points.add(Point(bounds.right, py))
                    }
                } else if (rightCheck and BoundsCheck.TOP != 0) {
                    py = bounds.top
                    if (newCheck and BoundsCheck.RIGHT != 0) {
                        px = bounds.right
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.LEFT != 0) {
                        px = bounds.left
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.BOTTOM != 0) {
                        if (rightPoint.x - bounds.x + newPoint.x - bounds.x < bounds.width) {
                            px = bounds.left
                        } else {
                            px = bounds.right
                        }
                        points.add(Point(px, py))
                        points.add(Point(px, bounds.bottom))
                    }
                } else if (rightCheck and BoundsCheck.BOTTOM != 0) {
                    py = bounds.bottom
                    if (newCheck and BoundsCheck.RIGHT != 0) {
                        px = bounds.right
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.LEFT != 0) {
                        px = bounds.left
                        points.add(Point(px, py))
                    } else if (newCheck and BoundsCheck.TOP != 0) {
                        if (rightPoint.x - bounds.x + newPoint.x - bounds.x < bounds.width) {
                            px = bounds.left
                        } else {
                            px = bounds.right
                        }
                        points.add(Point(px, py))
                        points.add(Point(px, bounds.top))
                    }
                }
            }
            if (closingUp) {
                // newEdge's ends have already been added
                return
            }
            points.add(newPoint!!)
        }
        val newRightPoint = newEdge.clippedEnds!![LR.other(newOrientation)]
        if (!closeEnough(points[0], newRightPoint!!)) {
            points.add(newRightPoint!!)
        }
    }

    val _x: Double
        get() = coord!!.x

    val _y: Double
        get() = coord!!.y

    fun dist(p: ICoord): Double {
        return Point.distance(p.coord!!, this.coord!!)
    }

    companion object {

        private val EPSILON = .005
        private val pool = Stack<Site>()

        fun create(p: Point, index: Int, weight: Double, color: Color?): Site {
            if (pool.size > 0) {
                return pool.pop().init(p, index, weight, color)
            } else {
                return Site(p, index, weight, color)
            }
        }

        fun sortSites(sites: ArrayList<Site>) {
            //sites.sort(Site.compare);
            Collections.sort(sites) { o1, o2 -> Site.compare(o1, o2).toInt() }
        }

        /**
         * sort sites on y, then x, coord also change each site's _siteIndex to
         * match its new position in the list so the _siteIndex can be used to
         * identify the site for nearest-neighbor queries
         *
         *
         * haha "also" - means more than one responsibility...
         */
        private fun compare(s1: Site, s2: Site): Double {
            val returnValue = Voronoi.compareByYThenX(s1, s2)

            // swap _siteIndex values if necessary to match new ordering:
            val tempIndex: Int
            if (returnValue == -1) {
                if (s1._siteIndex > s2._siteIndex) {
                    tempIndex = s1._siteIndex
                    s1._siteIndex = s2._siteIndex
                    s2._siteIndex = tempIndex
                }
            } else if (returnValue == 1) {
                if (s2._siteIndex > s1._siteIndex) {
                    tempIndex = s2._siteIndex
                    s2._siteIndex = s1._siteIndex
                    s1._siteIndex = tempIndex
                }

            }

            return returnValue.toDouble()
        }

        private fun closeEnough(p0: Point, p1: Point): Boolean {
            return Point.distance(p0, p1) < EPSILON
        }
    }
}
